### DISCUSSION TITLE
How next.js plans to implement data layer?

### DISCUSSION THREAD

**Discussion by kawpii** [vercel/next.js#15664] [Help]
With recent releases it seems Next.js is coming to SSG space. But I am wondering how to you guys plan to implement the data layer? ex: Gatsby's GraphQL. Data layer is important to separate data (ex: blog post content) from the template and to have easier integration with headless CMSes.

**Comment by stigkj** [comment]
I don't think they will do that. There are a lot of examples in the repo with how to access a datalayer through GraphQL, for example. And also examples with integration with CMSs.

**Reply by kawpii** [reply]
Can you please link those examples for Data Layer through GraphQL access, Integration with CMSes. So answer looks complete.

**Reply by stigkj** [reply]
Good call :-)

Examples of using a CMS for generating static content build time:
1. With [Cosmic](https://www.cosmicjs.com/) CMS: https://github.com/vercel/next.js/tree/canary/examples/cms-cosmic
2. With [GraphCMS](https://graphcms.com/) CMS: https://github.com/vercel/next.js/tree/canary/examples/cms-graphcms

Example 2 is actually using GraphQL for accessing the CMS data.

And last an example of using [GraphQL Hooks](https://github.com/nearform/graphql-hooks) from NearForm to fetch initial data build time: https://github.com/vercel/next.js/tree/canary/examples/with-graphql-hooks. This example also uses the regeneration feature that enables the static content to be refreshed every x seconds. See [`revalidate`](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation) for more info.

**Reply by kawpii** [reply]
Seems things have not stabilized yet since canary
